clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -858,6 +861,64 @@ class KernelObjVisitor {
   void VisitRecord(const CXXRecordDecl *Owner, ParentTy &Parent,
                    const CXXRecordDecl *Wrapper, Handlers &... handlers);
 
+  // Base case, only calls these when filtered.
+  template <typename... FilteredHandlers, typename ParentTy>
+  void VisitUnion(FilteredHandlers &... handlers, const CXXRecordDecl *Owner,
+                  ParentTy &Parent, const CXXRecordDecl *Wrapper) {
+    (void)std::initializer_list<int>{
+        (handlers.enterUnion(Owner, Parent), 0)...};
+    VisitRecordHelper(Wrapper, Wrapper->fields(), handlers...);
+    (void)std::initializer_list<int>{
+        (handlers.leaveUnion(Owner, Parent), 0)...};
+  }
+
+  // Handle empty base case.
+  template <typename ParentTy>
+  void VisitUnion(const CXXRecordDecl *Owner, ParentTy &Parent,
+                  const CXXRecordDecl *Wrapper) {}
+
+  // Add overloads including the base case when
+  // sizeof...(FilteredHandlers) > 0.
```

Comment:
In the version I sent you, I found we needed the sizeof...(FilteredHandlers) > 0 for The above case, and for the two with enable-ifs.  Did you miss those?  I don't think this will compile on all 3 compilers otherwise.

Reply:
sorry I missed this.

Reply:
Done.